### PR TITLE
Improve test coverage for modern girder_worker infrastructure

### PR DESCRIPTION
**Pull request by kotfic**
We can do better! Mostly hoping to improve test coverage of ```girder_worker/app.py``` and ```girder_worker/docker/*``` code. 

```
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
girder_worker/__init__.py                        20      0   100%
girder_worker/__main__.py                        15      3    80%
girder_worker/app.py                            217     84    61%
girder_worker/celeryconfig.py                    11      0   100%
girder_worker/configure.py                       44      1    98%
girder_worker/docker/__init__.py                  6      0   100%
girder_worker/docker/io/__init__.py             168     72    57%
girder_worker/docker/io/girder.py                15     15     0%
girder_worker/docker/nvidia.py                   52     52     0%
girder_worker/docker/stream_adapter.py           37     37     0%
girder_worker/docker/tasks/__init__.py          186    186     0%
girder_worker/docker/transforms/__init__.py     150    150     0%
girder_worker/docker/transforms/girder.py        43     43     0%
girder_worker/docker/utils.py                    46     46     0%
girder_worker/entrypoint.py                      78      5    94%
girder_worker/log_utils.py                       15      1    93%
girder_worker/utils.py                          105     41    61%
-----------------------------------------------------------------
TOTAL                                          1208    736    39%
```

This is currently based on https://github.com/girder/girder_worker/pull/225

### CODE REVIEW

**Comment by kotfic** [comment]
@cjh1 FYI

**Comment by kotfic** [comment]
```
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
girder_worker/__init__.py                        20      0   100%
girder_worker/__main__.py                        15      3    80%
girder_worker/app.py                            218     37    83%
girder_worker/celeryconfig.py                    11      0   100%
girder_worker/configure.py                       44      1    98%
girder_worker/docker/__init__.py                  6      0   100%
girder_worker/docker/io/__init__.py             168      7    96%
girder_worker/docker/io/girder.py                15      0   100%
girder_worker/docker/nvidia.py                   52     17    67%
girder_worker/docker/stream_adapter.py           37      1    97%
girder_worker/docker/tasks/__init__.py          186    156    16%
girder_worker/docker/transforms/__init__.py     150     11    93%
girder_worker/docker/transforms/girder.py        43      0   100%
girder_worker/docker/utils.py                    46     37    20%
girder_worker/entrypoint.py                      78      5    94%
girder_worker/log_utils.py                       15      1    93%
girder_worker/utils.py                          105     41    61%
-----------------------------------------------------------------
TOTAL                                          1209    317    74%
```

This could use significantly better coverage on ```docker/utils.py``` and  ```docker/tasks/__inti__.py```.  But those will have to wait until @cjh1 returns because they're significantly more complex. I think those can be addressed in a separate PR

**Comment by cjh1** [comment]
@kotfic The tests in ```task_plugin_test.py``` seem to be failing:

```
____________________________ TestTaskPlugin.test_core_plugin _____________________________

self = <tests.task_plugin_test.TestTaskPlugin testMethod=test_core_plugin>
app = <MagicMock name='app' id='139653155408848'>

    @mock.patch('girder_worker.__main__.app')
    def test_core_plugin(self, app):
>       main()

task_plugin_test.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../girder_worker/__main__.py:17: in main
    'CELERY_IMPORTS': get_core_task_modules()
../girder_worker/entrypoint.py:61: in get_core_task_modules
    return get_task_imports(get_extension_manager(app=app)['core'])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <stevedore.extension.ExtensionManager object at 0x7f0388b3c6d0>, name = 'core'

    def __getitem__(self, name):
        """Return the named extension.
    
            Accessing an ExtensionManager as a dictionary (``em['name']``)
            produces the :class:`Extension` instance with the
            specified name.
            """
>       return self._extensions_by_name[name]
E       KeyError: 'core'

/home/cjh/virtualenv/utest/local/lib/python2.7/site-packages/stevedore/extension.py:326: KeyError
```

Do I need to install the package or something to run these tests?

**Comment by jbeezley** [comment]
> Do I need to install the package or something to run these tests?

Yes, entrypoints only work in installed packages.  We could probably improve the error logging there though.

File: tests/test_app.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+def test_GirderAsyncResult_job_property_returns_None_if_no_jobs_found():
+    gar = GirderAsyncResult('BOGUS_TASK_ID')
+    with mock.patch('girder.utility.model_importer.ModelImporter') as mi:
+        mi.model.return_value.findOne.side_effect = IndexError()
+        assert gar.job is None
+
+
+def test_Task_AsynResult_of_type_GirderAsyncResult():
+    assert isinstance(Task().AsyncResult('BOGUS_TASK_ID'), GirderAsyncResult)
```
Comment (danlamanna) [inline tests/test_app.py]: Async*

Does it make sense to stub out a `pytest.mark.skip('unimplemented')` test here which is retrieving the `job` property when Girder isn't importable (on a remote worker)?

File: girder_worker/docker/nvidia.py [hunk 2]
```diff
@@ -38,7 +38,7 @@ def get_nvidia_configuration():
 
 def is_nvidia_image(api, image):
     labels = api.inspect_image(image).get('Config', {}).get('Labels')
-    return labels and labels.get('com.nvidia.volumes.needed') == 'nvidia_driver'
+    return bool(labels and labels.get('com.nvidia.volumes.needed') == 'nvidia_driver')
```
Comment (cjh1) [inline girder_worker/docker/nvidia.py]: Why is this needed? Isn't == always going to produce a bool?

Comment (kotfic) [reply girder_worker/docker/nvidia.py]: labels will be None if ```api.inspect_image(image)``` returns {'Config': {}},  and in a weird turn of events,  ```None and {True|False}```  returns None. womp. 

```python
In [1]: print(None and True)
None

In [2]: print(None and False)
None

In [3]: print(False and None)
False

In [4]: print(True and None)
None

```

Comment (cjh1) [reply girder_worker/docker/nvidia.py]: What about ```labels is not None``` ?

Comment (zachmullen) [reply girder_worker/docker/nvidia.py]: > in a weird turn of events, None and {True|False} returns None. womp.

In both python and javascript, `and` and `or` (`&&` and `||`) have the same behavior:

* `and`: returns the left value if it's falsy, otherwise returns the right value
* `or`: returns the left value if it's truthy, otherwise returns the right value

File: tests/test_docker_utils.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# TODO: test select_loop
+@pytest.mark.skip
+def test_select_loop():
+    pass
+
+
+# TODO; test chmod_writable
+@pytest.mark.skip
+def test_chmod_writable():
+    pass
```
Comment (cjh1) [inline tests/test_docker_utils.py]: I think we should remove these. Not sure there would be much value in the test as they would be pretty mocked out. They are pretty will covered by the integration tests.

File: tests/test_docker_stream_adapter.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            '\x01\x00\x00\x00\x00\x00\x00\x0chello world!'
+        ]
+
+        capture = CaptureAdapter()
+        adapter = DockerStreamPushAdapter(capture)
+        for d in data:
+            adapter.write(d)
+
+        self.assertEqual(capture.captured(),
+                         'this is stderr data\nthis is stdout data\nhello world!')
```
Comment (cjh1) [inline tests/test_docker_stream_adapter.py]: I assume this code was copied from the "old" test, can we remove the code from the old test?
